index.bs

```diff
@@ -4046,7 +4073,59 @@ To override this default policy and indicate that a cross-origin <{iframe}> is a
+
+[=[WRPS]=] can opt in to allowing [=WebAuthn Clients=] to enable a credential to be created and used across a limited set of related [=origin|origins=]
+
+Relying Parties MUST choose a common [=RP ID=] to use across all ceremonies from related origins.
 
+A JSON document MUST be hosted at the `webauthn` well-known URL ([[!RFC8615]]) for the [=RP ID=]. The JSON document MUST be returned as follows:
+
+    - The content type MUST be `application/json`.
+    - JSON document MUST contain a single object named `origins`.
+    - The `origins` object MUST contain an array of one or more web origins.
+
+For example, for the RP ID `example.com`:
+
+<xmp class="json" highlight="json">
+{
+    "origins": [
+        "https://example.co.uk",
+        "https://example.de",
+        "https://example.sg"
+        "https://myexamplerewards.com"
+    ]
+}
+</xmp>
+
+[=WebAuthn Clients=] supporting this feature MUST support at least five [=origin labels=]. Client policy SHOULD define an upper limit to prevent abuse.
+
+[=WebAuthn Clients=] supporting this feature SHOULD include {{ClientCapability/relatedOrigins}} in their response to [[#sctn-getClientCapabilities|getClientCapabilities()]].
+
+### Validating Related Origins ### {#sctn-validating-relation-origin}
+
+To validate the calling origin is an authorized related origin for a given ceremony:
+
+1. Let |maxLabels| be the maximum number of [=origin labels=] allowed by client policy.
+1. Fetch the well-known URL for the RP ID (<code>https://|rpIdRequested|/.well-known/webauthn</code>).
+    1. If the fetch fails, the response does not have a content type of `application/json`, or does not have a status code (after following redirects) of 200, then throw a "{{SecurityError}}" {{DOMException}}.
+    1. If the body of the resource is not a valid JSON object, then throw a "{{SecurityError}}" {{DOMException}}.
+    1. If the value of the |origins| property of the JSON object is missing, or is not an array of strings, then throw a "{{SecurityError}}" {{DOMException}}.
+1. Let |labelsSeen| be a new empty [=set=].
+1. [=set/For each=] string in |origins|:
+    1. Let |url| be the result of parsing the string as a URL. If that fails, [=continue=].
+    1. Let |domain| be the [=effective domain=] of |url|. If that is null, [=continue=].
+    1. Remove any [=public suffix=] from the end of |domain|, including private registries and unknown registries. If |domain| is now empty, [=continue=].
+    1. Split |domain| into [=origin labels|labels=] and let |label| be the right-most one.
+    1. If |labelsSeen| [=set/contains=] |label|, [=continue=].
+    1. If |rpIdRequested| and |url| are [=same origin=], return [TRUE].
+    1. [=set/Append=] |label| to |labelsSeen|.
+    1. If the [=set/size=] of |labelsSeen| is greater than or equal to |maxLabels|, [=break=].
```

Comment:
On second read I think there's a wrinkle in this logic, namely that it returns immediately as soon as the 5th (say) _labelsSeen_ element is added, so the "If _labelsSeen_ contains _label_, continue" condition doesn't get a chance to trigger for _domain_ values past that. I.e., the origin `https://www.foo1.org` would fail with this related origin list:

- `https://foo1.org`
- `https://foo2.org`
- `https://foo3.org`
- `https://foo4.org`
- `https://foo5.org` (_labelsSeen_ is now `["foo1", "foo2", "foo3", "foo4", "foo5"]`, break and return `false`)
- `https://www.foo1.org` (_labelsSeen_ is still only 5 elements, so procedure would return `true` here)

I think this could be fixed by simply changing "greater than or equal to" to "greater than", if that is closer to the intent?

Reply:
@emlun did a quick test with this code and I think `>=` is correct.

```javascript
var psl = require('psl');

const MAX_LABELS = 5;

//const WELL_KNOWN = {
//    origins: [
//        "https://shopping.sg",
//        "https://shopping.ie",
//        "https://shopping.co.uk",
//        "https://otherdomain1.com",
//        "https://otherdomain2.com",
//        "https://otherdomain3.com",
//        "https://otherdomain4.com",
//        "https://otherdomain5.com"
//    ]
//}

const WELL_KNOWN = {
    origins: [
        "https://otherdomain1.com",
        "https://otherdomain2.com",
        "https://otherdomain3.com",
        "https://otherdomain4.com",
        "https://otherdomain5.com",
        "https://otherdomain6.com"
    ]
}

function checkRpId(rpIdRequested) {
    const labelsSeen = new Set();

    for (const origin of WELL_KNOWN.origins) {
        const url = new URL(origin);

        let domain = url.hostname;

        if (!domain) continue;

        // grab eTLD+1 label
        const parsed = psl.parse(domain)
        const label = parsed.sld;

        if (labelsSeen.has(label)) continue;

        if (`https://${rpIdRequested}` === url.origin) return true;

        labelsSeen.add(label);

        if (labelsSeen.size >= MAX_LABELS) break;
    }
    return false;
}
```

Examples:

```javascript
console.log(checkRpId("otherdomain6.com"))
```
`false`

<hr>

```javascript
console.log(checkRpId("otherdomain5.com"))
```
`true`

Reply:
@MasterKale please sanity check ^

Reply:
> If the [=set/size=] of |labelsSeen| is greater than or equal to |maxLabels|, [=break=].

If anything for sake of readability I'd put this check at the top of the list of steps. Consider the following amendment to the code above:

From this...
```js
for (const origin of WELL_KNOWN.origins) {
    // ...
    if (labelsSeen.size >= MAX_LABELS) break;
}
```
...to this:
```js
for (const origin of WELL_KNOWN.origins) {
    if (labelsSeen.size === MAX_LABELS) break;
    // ...
}
```

The latter leaves less room for edge cases as it more explicitly communicates the fact that only up to five related labels are supported.

Reply:
I'm still considering Emil's question, one sec...

Reply:
There appears to be a bug in this logic. RP ID scoping rules allow for RP ID `otherdomain5.com` to be specified if WebAuthn is handled from `https://www.otherdomain5.com`.

However, if I call `checkRpId("otherdomain5.com")` as written above, with the following list of related origins...

```js
let WELL_KNOWN = {
    origins: [
        "https://otherdomain1.com",
        "https://otherdomain2.com",
        "https://otherdomain3.com",
        "https://otherdomain4.com",
        "https://www.otherdomain5.com",
    ]
}
```

...the method return `false`. This list is still within the client's maximum of five labels, however I think this step is part of the problem:

> If |rpIdRequested| and |url| are [=same origin=], return [TRUE].

This likely needs to get tweaked to say something like, "rpIdRequested is valid for the current URL" or some such 🤔

Reply:
> There appears to be a bug in this logic. RP ID scoping rules allow for RP ID otherdomain5.com to be specified if WebAuthn is handled from https://www.otherdomain5.com.

Related Origins logic would not be called in this case, as this would pass the existing RP < > origin check in the spec.

Reply:
I forget about the checks that must occur before this logic is evaluated:

![Screenshot 2024-06-14 at 10 23 29 AM](https://github.com/w3c/webauthn/assets/5166470/7bedde80-0dbb-4605-b0e3-22b382a4dc5e)